synthesizer/process/src/lib.rs

```diff
@@ -205,26 +250,74 @@ impl<N: Network> Process<N> {
     /// Returns `true` if the process contains the program with the given ID.
     #[inline]
     pub fn contains_program(&self, program_id: &ProgramID<N>) -> bool {
-        self.stacks.contains_key(program_id)
+        if self.credits.as_ref().map_or(false, |stack| stack.program_id() == program_id) {
+            return true;
+        }
+        let lock = self.stacks.lock();
+        match lock {
+            Ok(lock) => lock.contains(program_id),
+            Err(_) => false,
+        }
+    }
+
+    fn load_stack(&self, program_id: impl TryInto<ProgramID<N>>) -> Result<()> {
+        let program_id = program_id.try_into().map_err(|_| anyhow!("Invalid program ID"))?;
+        info!("Lazy loading stack for {program_id}");
+        let storage_mode = self.storage_mode.as_ref().ok_or_else(|| anyhow!("Failed to get storage mode"))?.clone();
+        // try to lazy load the stack
+        #[cfg(feature = "rocks")]
+        let store = ConsensusStore::<N, ConsensusDB<N>>::open(storage_mode);
+        #[cfg(not(feature = "rocks"))]
+        let store = ConsensusStore::<N, ConsensusMemory<N>>::open(storage_mode);
```

Comment:
It may be easier to either have `C: ConsensusStorage` as a generic or to pass in `storage` somehow. 

It might break some APIs, but it feels cleaner than reloading the storage it at each instance.

Reply:
yes, this was the original idea. but adding the generic cascaded everywhere... Will look into that.

Reply:
This cascades too much. The perf hit to open storage is about 2s (on a 15GB ledger). That's not negligible, but hopefully we don't need to keep so many programs in memory.

Reply:
found a way to avoid opening storage every time :tada: